src/fontra/views/editor/edit-tools-shape.js

```diff
@@ -0,0 +1,188 @@
+
+  async handleDrag(eventStream, initialEvent) {
+    if (!this.sceneModel.selectedGlyph?.isEditing) {
+      await this.editor.tools["pointer-tool"].handleDrag(eventStream, initialEvent);
+      return;
+    }
+
+    const initialPoint = this.sceneController.selectedGlyphPoint(initialEvent);
+
+    if (!(await shouldInitiateDrag(eventStream, initialEvent))) {
+      // TODO: open dialog for numeric size input
+      return;
+    }
+
+    let mouseRect;
+    let eventTemp;
+    for await (const event of eventStream) {
+      eventTemp = event;
+      const point = this.sceneController.selectedGlyphPoint(event);
+      if (point.x === undefined) {
+        // We can receive non-pointer events like keyboard events: ignore
+        continue;
+      }
+      mouseRect = rectangle.rectRound({
+        xMin: initialPoint.x,
+        yMin: initialPoint.y,
+        xMax: point.x,
+        yMax: point.y,
+      });
+
+      const drawPath = new Path2D();
+      this.drawShapeWithKeys(drawPath, mouseRect, event);
+      this.sceneModel.shapeToolShapePath = drawPath;
+      this.sceneModel.event = event;
+      this.canvasController.requestUpdate();
+    }
+
+    delete this.sceneModel.shapeToolShapePath;
+    delete this.sceneModel.event;
+    this.canvasController.requestUpdate();
+
+    // rectsize return when too small
+    if (!mouseRect) {
+      return;
+    }
+
+    const pathNew = new VarPackedPath();
+    this.drawShapeWithKeys(pathNew, mouseRect, eventTemp);
+    this.addShapePath(pathNew);
+  }
+
+  async addShapePath(pathNew) {
+    await this.sceneController.editGlyphAndRecordChanges(
+      (glyph) => {
+        const editLayerGlyphs = this.sceneController.getEditingLayerFromGlyphLayers(
+          glyph.layers
+        );
+
+        const firstLayerGlyph = Object.values(editLayerGlyphs)[0];
+        const selection = new Set();
+        const firstIndex = firstLayerGlyph.path.numPoints;
+        for (const index of range(pathNew.numPoints)) {
+          const point = pathNew.getPoint(index);
+          if (!point.type) {
+            selection.add(`point/${firstIndex + index}`);
+          }
+        }
+        this.sceneController.selection = selection;
+
+        for (const [layerName, layerGlyph] of Object.entries(editLayerGlyphs)) {
+          layerGlyph.path.appendPath(pathNew);
+        }
+        return "add shape";
+      },
+      undefined,
+      true
+    );
+  }
+
+  deactivate() {
+    this.canvasController.requestUpdate();
+  }
+
+  drawShapeWithKeys(path, mouseRect, event) {
+    let x = mouseRect.xMin;
+    let y = mouseRect.yMin;
+    let width = mouseRect.yMax - mouseRect.yMin;
+    let height = mouseRect.xMax - mouseRect.xMin;
+
+    if (event.shiftKey) {
+      // make square, not rectangle
+      if ((width > 0 && height > 0) || (width < 0 && height < 0)) {
+        height = width;
+      } else {
+        height = -width;
+      }
+    }
+
+    let revesed = event.ctrlKey ? true : false; // reversed contour direction
```

Comment:
```suggestion
    let reversed = event.ctrlKey; // reversed contour direction
```
This property is already a boolean. Same for the next line.

Reply:
At the same time, I think all flags should be handled here, like the shiftKey is handled here.

Define drawShapePath like this:
```js
    drawShapePath(path, x, y, width, height) {
```
And feed it the "right" arguments based on the modifiers.

Reverse also doesn't need to be special cased. There are two ways to do it:
- reverse the path after it is constructed
- use a negative width or height, adjust x or y

Reply:
I agree. As mentioned yesterday, I had some issues with that. I started to add a reverse method, but as we have to deal with two different objects (path2D and VarPackedPath) the same way, I was not able to make it work yesterday, why I ended up with this solution. As said: not nice, but it works.

Reply:
For reverence, reversePath vs. flip:
<img width="1718" alt="Screenshot 2024-03-20 at 14 26 29" src="https://github.com/googlefonts/fontra/assets/21055547/8b4a4af9-217f-4566-90d1-e96f995ffd0e">
Different start point location.